### ISSUE TITLE
A couple of notes about the new edit meta-data table.

### ISSUE THREAD

**Issue by ljchang** [NeuroVault/NeuroVault#317]
@burnash @chrisfilo @vsoch Nice job with this new function!  Overall, it is working well, particularly for smaller collections.  I just wanted to provide a couple of issues I've run into before I forget.

1) It would be nice if there was a button to download the table.

2) There is no way to bulk delete data.   I have received errors for repeated data, but there was no way to delete the data that was repeated.  I had to manually go to the images and delete each one in a separate window, which is very slow and tedious.  Could we add a delete row/image button?  Or might this cause too many problems?  

3) When working on a very large table (~5k rows), it is very slow to add columns (takes several minutes), also I have received a gateway time-out error upon validation.  However, it looks like the data is being saved correctly despite this warning.  Here's a screenshot.

![2015-08-10_12-02-15](https://cloud.githubusercontent.com/assets/2993338/9176640/7551d9ac-3f5a-11e5-8f37-11b1a49a83e3.png)

4) It would be nice to double click in the bottom right corner of an element to repeat, similar to excel/google doc.  This works when the column is empty and preceded by a separate column, but not when the column on the left is empty, or if the column is already populated with another value.

I'll post more examples of my experience as issues arise.

**Comment by chrisfilo** [comment]
On Mon, Aug 10, 2015 at 9:26 AM, Luke Chang <notifications@github.com>
wrote:

> @burnash <https://github.com/burnash> @chrisfilo
> <https://github.com/chrisfilo> @vsoch <https://github.com/vsoch> Nice job
> with this new function! Overall, it is working well, particularly for
> smaller collections. I just wanted to provide a couple of issues I've run
> into before I forget.
>
> 1) It would be nice if there was a button to download the table.
>
I believe @burnash is working on adding this in the next release.

> 2) There is no way to bulk delete data. I have received errors for
> repeated data, but there was no way to delete the data that was repeated. I
> had to manually go to the images and delete each one in a separate window,
> which is very slow and tedious. Could we add a delete row/image button? Or
> might this cause too many problems?
>
This can be a bit confusing. If we allow deleting rows there needs to be a
clear dialog box saying that this will lead to irreversible data loss.

> 3) When working on a very large table (~5k rows), it is very slow to add
> columns (takes several minutes), also I have received a gateway time-out
> error upon validation. However, it looks like the data is being saved
> correctly despite this warning. Here's a screenshot.
>
> [image: 2015-08-10_12-02-15]
> <https://cloud.githubusercontent.com/assets/2993338/9176640/7551d9ac-3f5a-11e5-8f37-11b1a49a83e3.png>
>
> 4) It would be nice to double click in the bottom right corner of an
> element to repeat, similar to excel/google doc. This works when the column
> is empty and preceded by a separate column, but not when the column on the
> left is empty, or if the column is already populated with another value.
>
 I was looking for this an it actually works. If you select a cell and then
click on it again a square appears in the right corner of the cell - you
can pull it down to repeat the content of it in other cells.

**Comment by burnash** [comment]
> 1) It would be nice if there was a button to download the table.

@chrisfilo's right. I have a similar feature in the csv import PR. It'll be quite straightforward to reimplement it on the edit metadata page. If nobody has any objections, we can put the export button in the same place as in CogatPheno's table:

![export button](https://cloud.githubusercontent.com/assets/814322/8938773/800ae6bc-3515-11e5-92ce-90262b05b4f1.png)
<br/>

> 2) There is no way to bulk delete data. I have received errors for repeated data, but there was no way to delete the data that was repeated. I had to manually go to the images and delete each one in a separate window, which is very slow and tedious. Could we add a delete row/image button? Or might this cause too many problems?

I'm sure both @vsoch and @chrisfilo have a deeper understanding of the consequences of images deletion than I do. UI-wise it can be easily done in similar manner to column deletion (with context menu.) Adding a confirmation modal dialog is also possible (same as in the add column case.) I didn't test row deletion in huge handsontable containers though.

> 3) When working on a very large table (~5k rows), it is very slow to add columns (takes several minutes), also I have received a gateway time-out error upon validation. However, it looks like the data is being saved correctly despite this warning. Here's a screenshot.

I'm still investigating the source of the slowness. The debugger points to the Handsontable's `render` method which I call to redraw the table with a newly inserted column. I didn't profile the code thoroughly yet. The timeout error probably comes from Nginx. The validation for 5k rows may run longer than Nginx expects from the uwsgi server (it is a `uwsgi_read_timeout` parameter in the config, [which defaults to 60s](http://nginx.org/en/docs/http/ngx_http_uwsgi_module.html#uwsgi_read_timeout))

> 4) It would be nice to double click in the bottom right corner of an element to repeat, similar to excel/google doc. This works when the column is empty and preceded by a separate column, but not when the column on the left is empty, or if the column is already populated with another value.

@chrisfilo is correct. After you enter a new value you need to click the cell again for a little blue square in the right bottom corner to reappear.

**Comment by vsoch** [comment]
Our current "[delete image](https://github.com/NeuroVault/NeuroVault/blob/master/neurovault/apps/statmaps/views.py#L574)" functionality is intended for one image, as it takes the primary key in the url

       http://neurovault.org/images/299/delete

Could you better describe the use case for deleting images from the metadata table, as opposed to bulk delete from the images datatable? I could imagine having the ability to select multiple from the main image table, but I don't like the idea of deleting a row in the metadata --> deleting the image. If it is the case that all images are required in order to add metadata, perhaps that should be waived. I do like the idea of having functions to work with large data in the meta table, for example highlighting / deleting repeated rows (and NOT deleting the associated image).

For the actual process of image deletion - given that we allow users to select multiple images from the main datatable, how about something like the following:

1.  click some "delete multiple" button that exposes check marks next to images. The user can "select all" or "deselect all" (and it should only be for those visible on the screen, given pagination). This also exposes a "delete" button.
2. after selection, clicking "delete" gives a final warning about data loss, etc.
3. On the back end, we send the list of ids to the server.  We could either create a temporary faux collection, move the images to it, and delete the collection, or we could do some loop and delete them one at a time. My thinking is that the first might be faster, because Django is good at that sort of thing.

The Handsontable with 5000 rows shouldn't be slow - take a [look here](https://github.com/handsontable/handsontable/wiki/FAQ#handsontable-seems-slow). I haven't looked at how you have it configured, but could we select a constrained viewport or specify column widths and see if that speeds things up a bit?

**Comment by ljchang** [comment]
Thanks for your thoughts.  Dragging the corner of the blue box definitely works and is very handy.  You can also double click the corner to repeat the rows if they are empty, just like excel.  However, this function won't override data if the rows are already populated.  It's possible this is just for specific types of data (e.g., map_type).  Also, the double clicking does not work if the column to the left of the active column is empty for some reason.  Excel does this too, which I've never quite understood.  To be honest, I think this functionality is good enough for now and don't think it's worth putting any more time into it.

The bulk delete functionality however is something that would be very useful to have somewhere on the app.  @vsoch I didn't quite see any bulk delete options except for the entire collection.  This is of course useful, but it's very possible that users want to delete multiple images, but not the entire collection.  They shouldn't have to view each image individually to do this.  I think having a 'select/unselect' data checkbox and a 'delete selected' button with confirmation is something that would be very useful (it might also be nice to have a 'download selected' button as well).  I agree it might make more sense to have it on the image data table rather than on the metadata table, which is a bit obscure.  If we go this route I agree that we should remove the duplicate image validation.  Both of those suggested delete methods on the backend seem reasonable.  With the pagination, I can't imagine that it would take much longer to delete the images selected on the screen.

**Comment by vsoch** [comment]
@ljchang I completely agree - I've needed bulk delete many times (in dev) and usually resort to deleting the collection. It looks like we've talked about this before https://github.com/NeuroVault/NeuroVault/issues/120 and it's just a matter of getting it into a PR!  My [current PR](https://github.com/NeuroVault/NeuroVault/pull/315) needs a few final tweaks, but I think I could start work on multiple image operations soon.
